atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/AbstractDegradedClusterTest.java

```diff
@@ -0,0 +1,127 @@
+import static org.assertj.core.api.Assertions.fail;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.thrift.TException;
+import org.junit.Before;
+
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraClient;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueService;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServices;
+import com.palantir.common.base.RunnableCheckedException;
+import com.palantir.common.exception.AtlasDbDependencyException;
+
+public abstract class AbstractDegradedClusterTest {
+    static final TableReference TEST_TABLE = TableReference.createWithEmptyNamespace("test_table");
+    static final byte[] FIRST_ROW = PtBytes.toBytes("row1");
+    static final byte[] SECOND_ROW = PtBytes.toBytes("row2");
+    static final byte[] FIRST_COLUMN = PtBytes.toBytes("col1");
+    static final byte[] SECOND_COLUMN = PtBytes.toBytes("col2");
+    static final Cell CELL_1_1 = Cell.create(FIRST_ROW, FIRST_COLUMN);
+    static final Cell CELL_1_2 = Cell.create(FIRST_ROW, SECOND_COLUMN);
+    static final Cell CELL_2_1 = Cell.create(SECOND_ROW, FIRST_COLUMN);
+    static final Cell CELL_2_2 = Cell.create(SECOND_ROW, SECOND_COLUMN);
+    static final byte[] CONTENTS = PtBytes.toBytes("default_value");
+    static final long TIMESTAMP = 2L;
+    static final Value VALUE = Value.create(CONTENTS, TIMESTAMP);
+
+    private static final Map<Class<? extends AbstractDegradedClusterTest>, CassandraKeyValueService> testKvs =
+            new HashMap<>();
+
+    private String schemaAtStart;
+
+    @Before
+    public void recordSchemaVersion() throws TException {
+        schemaAtStart = getUniqueReachableSchemaVersionOrThrow();
+    }
+
+    /**
+     * Used in {@link NodesDownTestSetup#initializeKvsAndDegradeCluster(List, List)} using reflection to perform any
+     * necessary initialization before degrading the cluster.
+     *
+     * @param kvs a dedicated instance of CKVs keyed to a specific namespace to be used with this test only
+     */
+    public void initialize(CassandraKeyValueService kvs) {
+        testKvs.put(getClass(), kvs);
+        testSetup(kvs);
+    }
+
+    abstract void testSetup(CassandraKeyValueService kvs);
+
+    CassandraKeyValueService getTestKvs() {
+        return testKvs.get(getClass());
+    }
+
+    static void closeAll() {
+        testKvs.values().forEach(KeyValueService::close);
+        testKvs.clear();
+    }
+
+    void assertKvsReturnsGenericMetadata(TableReference tableRef) {
```

Comment:
Non-blocking: To me at least this seems like a strange place to park the assertions, though OK for now. I get that there is some logic and they are used by multiple tests, so a fluent kvs-assert could be interesting. That said this PR is already too long :)

Reply:
Yes, let's not do that in this pr 🗡